diff --git a/uarchsim/fetchunit.cc b/uarchsim/fetchunit.cc
index a8b1281..1594601 100644
--- a/uarchsim/fetchunit.cc
+++ b/uarchsim/fetchunit.cc
@@ -136,6 +136,12 @@ void fetchunit_t::transfer_fetch_bundle() {
       PAY->buf[index].branch_type = fetch_bundle[pos].branch_type;
       PAY->buf[index].branch_target = fetch_bundle[pos].branch_target;
       PAY->buf[index].fflags = 0; // fflags field is always cleaned for newly fetched instructions
+ 
+      // CPR: Initialize chkpt_id to something greater than the largest valid chkpt_id.
+      // If you don't do this, the retirement unit could prematurely pop instructions
+      // from PAY (during RETIRE_FINALIZE) that haven't passed RENAME2 yet, because their
+      // uninitialized chkpt_id could wrongly match RETSTATE.chkpt_id.
+      PAY->buf[index].chkpt_id = 0xDEADBEEF;
 
       // Clear the trap storage before the first time it is used.
       PAY->buf[index].trap.clear();
